### PR TITLE
feat(refunds): Refunds aggregate api

### LINKED ISSUE #5796: feat(payments): show aggregate count of refund status

### PR DESCRIPTION
**Pull request by gitanjli525**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
Added new api for refunds status aggregate

### Additional Changes

- [x] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context
Closed #5796 

## How did you test it?

Request
```
curl --location 'http://localhost:8080/refunds/aggregate?start_time=2022-08-28T18%3A30%3A00Z' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoiZjNmYzEyNjYtZDQ2MS00NWU3LWFkMzEtODdlMDA4ZWVlYjZjIiwibWVyY2hhbnRfaWQiOiJtZXJjaGFudF8xNzI0NjYxNjQ3Iiwicm9sZV9pZCI6Im9yZ19hZG1pbiIsImV4cCI6MTcyNTYyMDQ1MSwib3JnX2lkIjoib3JnX3pXYVBBZDBxOGtxcnRpQzN4U3o3IiwicHJvZmlsZV9pZCI6bnVsbH0.VXacvv96BOspj033EFwKvaYdyqjlSNIzMFXYQ-DvsvE' \
--header 'Cookie: login_token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoiZjNmYzEyNjYtZDQ2MS00NWU3LWFkMzEtODdlMDA4ZWVlYjZjIiwibWVyY2hhbnRfaWQiOiJtZXJjaGFudF8xNzI0NjYxNjQ3Iiwicm9sZV9pZCI6Im9yZ19hZG1pbiIsImV4cCI6MTcyNTYyMDQ1MSwib3JnX2lkIjoib3JnX3pXYVBBZDBxOGtxcnRpQzN4U3o3IiwicHJvZmlsZV9pZCI6bnVsbH0.VXacvv96BOspj033EFwKvaYdyqjlSNIzMFXYQ-DvsvE'
``` 

Response
```
{
    "status_with_count": {
        "pending": 0,
        "transaction_failure": 0,
        "success": 40,
        "failure": 0,
        "manual_review": 0
    }
}
``` 

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/api_models/src/refunds.rs [hunk 1]
```diff
@@ -234,6 +234,12 @@ pub struct RefundListFilters {
     pub refund_status: Vec<enums::RefundStatus>,
 }
 
+#[derive(Clone, Debug, serde::Serialize)]
```
Comment (apoorvdixit88) [inline crates/api_models/src/refunds.rs:237]: Please add `ToSchema` for open_api_spec

File: crates/router/src/core/refunds.rs [hunk 2]
```diff
@@ -1034,6 +1034,30 @@ pub async fn get_filters_for_refunds(
     ))
 }
```
Comment (apoorvdixit88) [inline crates/router/src/core/refunds.rs:1036]: Please add `#[instrument(skip_all)]`

File: crates/router/src/db/kafka_store.rs [hunk 3]
```diff
@@ -2389,6 +2389,18 @@ impl RefundInterface for KafkaStore {
             .await
     }
 
+    #[cfg(feature = "olap")]
+    async fn get_refund_status_with_count(
+        &self,
+        merchant_id: &id_type::MerchantId,
+        constraints: &api_models::payments::TimeRange,
+        storage_scheme: MerchantStorageScheme,
+    ) -> error_stack::Result<Vec<(common_enums::RefundStatus, i64)>, errors::StorageError> {
```
Comment (apoorvdixit88) [inline crates/router/src/db/kafka_store.rs:2398]: Why not `CustomResult`?

File: crates/router/src/db/refund.rs [hunk 4]
```diff
@@ -250,6 +258,20 @@ mod storage {
...
             .map_err(|error|report!(errors::StorageError::from(error)))
         }
+        #[cfg(feature = "olap")]
+        #[instrument(skip_all)]
+        async fn get_refund_status_with_count(
+            &self,
+            merchant_id: &common_utils::id_type::MerchantId,
+            time_range: &api_models::payments::TimeRange,
+            _storage_scheme: enums::MerchantStorageScheme,
+        ) -> error_stack::Result<Vec<(common_enums::RefundStatus, i64)>, errors::StorageError>
```
Comment (apoorvdixit88) [inline crates/router/src/db/refund.rs:268]: `CustomResult`*

File: crates/router/src/db/refund.rs [hunk 5]
```diff
@@ -250,6 +258,20 @@ mod storage {
...
+        {
+            let conn = connection::pg_connection_read(self).await?;
+            <diesel_models::refund::Refund as storage_types::RefundDbExt>::get_refund_status_with_count(&conn, merchant_id, time_range)
+            .await
```
Comment (apoorvdixit88) [inline crates/router/src/db/refund.rs:272]: `.map_err(|error|report!(errors::StorageError::from(error)))`

File: crates/router/src/db/refund.rs [hunk 6]
```diff
@@ -783,6 +805,19 @@ mod storage {
                         .map_err(|error|report!(errors::StorageError::from(error)))
         }
 
+        #[cfg(feature = "olap")]
```
Comment (apoorvdixit88) [inline crates/router/src/db/refund.rs:808]: Please add ` #[instrument(skip_all)]`

File: crates/router/src/db/refund.rs [hunk 7]
```diff
@@ -783,6 +805,19 @@ mod storage {
...
+        async fn get_refund_status_with_count(
+            &self,
+            merchant_id: &common_utils::id_type::MerchantId,
+            constraints: &api_models::payments::TimeRange,
+            _storage_scheme: enums::MerchantStorageScheme,
+        ) -> error_stack::Result<Vec<(common_enums::RefundStatus, i64)>, errors::StorageError>
```
Comment (apoorvdixit88) [inline crates/router/src/db/refund.rs:814]: use `CustomResult` *

File: crates/router/src/db/refund.rs [hunk 8]
```diff
@@ -783,6 +805,19 @@ mod storage {
...
+        {
+            let conn = connection::pg_connection_read(self).await?;
+            <diesel_models::refund::Refund as storage_types::RefundDbExt>::get_refund_status_with_count(&conn, merchant_id, constraints)
+            .await
```
Comment (apoorvdixit88) [inline crates/router/src/db/refund.rs:818]: .map_err(|error| report!(errors::StorageError::from(error)))

File: crates/router/src/routes/refunds.rs [hunk 9]
```diff
@@ -421,6 +423,36 @@ pub async fn get_refunds_filters_profile(
     .await
 }
 
+#[instrument(skip_all, fields(flow = ?Flow::RefundsAggregate))]
+#[cfg(feature = "olap")]
+pub async fn get_refunds_aggregates(
+    state: web::Data<AppState>,
+    req: HttpRequest,
+    payload: web::Query<payment_types::TimeRange>,
```
Comment (apoorvdixit88) [inline crates/router/src/routes/refunds.rs:431]: We can remove this import and qualify it as `api_models::payments::TimeRange`

File: crates/router/src/types/api/refunds.rs [hunk 10]
```diff
@@ -1,6 +1,6 @@
 pub use api_models::refunds::{
-    RefundRequest, RefundResponse, RefundStatus, RefundType, RefundUpdateRequest,
-    RefundsRetrieveRequest,
+    RefundAggregateResponse, RefundRequest, RefundResponse, RefundStatus, RefundType,
```
Comment (apoorvdixit88) [inline crates/router/src/types/api/refunds.rs:2]: Why this import?

File: crates/router/src/types/storage/refund.rs [hunk 11]
```diff
@@ -13,35 +13,41 @@ use diesel_models::{
...
+        conn: &connection::PgPooledConn,
         merchant_id: &common_utils::id_type::MerchantId,
         refund_list_details: &api_models::refunds::RefundListRequest,
     ) -> CustomResult<i64, errors::DatabaseError>;
+
+    async fn get_refund_status_with_count(
+        conn: &connection::PgPooledConn,
+        merchant_id: &common_utils::id_type::MerchantId,
+        time_range: &api_models::payments::TimeRange,
+    ) -> CustomResult<Vec<(RefundStatus, i64)>, StorageError>;
```
Comment (apoorvdixit88) [inline crates/router/src/types/storage/refund.rs:44]: Why not `errors::DatabaseError` here?

File: crates/router/src/types/storage/refund.rs [hunk 12]
```diff
@@ -288,4 +294,38 @@ impl RefundDbExt for Refund {
             .change_context(errors::DatabaseError::NotFound)
             .attach_printable_lazy(|| "Error filtering count of refunds")
     }
+
+    async fn get_refund_status_with_count(
+        conn: &connection::PgPooledConn,
+        merchant_id: &common_utils::id_type::MerchantId,
+        time_range: &api_models::payments::TimeRange,
+    ) -> error_stack::Result<Vec<(RefundStatus, i64)>, StorageError> {
```
Comment (apoorvdixit88) [inline crates/router/src/types/storage/refund.rs:302]: `CustomResult` with `errors::DatabaseError`

File: crates/router/src/types/storage/refund.rs [hunk 13]
```diff
@@ -288,4 +294,38 @@ impl RefundDbExt for Refund {
...
+        query = match time_range.end_time {
+            Some(ending_at) => query.filter(dsl::created_at.le(ending_at)),
+            None => query,
+        };
+
+        logger::debug!(filter = %diesel::debug_query::<diesel::pg::Pg,_>(&query).to_string());
+
+        db_metrics::track_database_call::<<Self as HasTable>::Table, _, _>(
+            query.get_results_async::<(RefundStatus, i64)>(conn),
+            db_metrics::DatabaseOperation::Filter,
```
Comment (apoorvdixit88) [inline crates/router/src/types/storage/refund.rs:320]: db_metrics::DatabaseOperation::Count

**Review by apoorvdixit88 — CHANGES_REQUESTED** [review]
Please fix above comments

File: crates/router/src/types/api/refunds.rs (hunk 10 above)
Comment (gitanjli525) [reply crates/router/src/types/api/refunds.rs:2]: using RefundAggregateResponse in core as `api::RefundAggregateResponse`, updated it to `api_models::refunds::RefundAggregateResponse`

**Comment by gitanjli525** [comment]
> Please fix above comments

updated the PR

File: crates/router/src/core/refunds.rs [hunk 14]
```diff
@@ -1034,6 +1034,31 @@ pub async fn get_filters_for_refunds(
...
+pub async fn get_aggregates_for_refunds(
+    state: SessionState,
+    merchant: domain::MerchantAccount,
+    time_range: api::TimeRange,
+) -> RouterResponse<api_models::refunds::RefundAggregateResponse> {
+    let db = state.store.as_ref();
+    let refund_status_with_count = db
+        .get_refund_status_with_count(merchant.get_id(), &time_range, merchant.storage_scheme)
+        .await
+        .to_not_found_response(errors::ApiErrorResponse::RefundNotFound)?;
```
Comment (apoorvdixit88) [inline crates/router/src/core/refunds.rs:1048]: Wont it be `InternalServerError`?

File: crates/router/src/routes/refunds.rs [hunk 15]
```diff
@@ -3,6 +3,7 @@ use common_enums::EntityType;
 use router_env::{instrument, tracing, Flow};
 
 use super::app::AppState;
+#[cfg(feature = "olap")]
```
Comment (apoorvdixit88) [inline crates/router/src/routes/refunds.rs:6]: Move this feature flag, use the feature flag only for olap import.

Comment (apoorvdixit88) [inline crates/router/src/routes/refunds.rs:6]: Move this feature flag, use the feature flag only for required import type

**Review by apoorvdixit88 — DISMISSED** [review]

File: crates/router/src/core/refunds.rs [hunk 16]
```diff
@@ -1034,6 +1034,32 @@ pub async fn get_filters_for_refunds(
...
+    state: SessionState,
+    merchant: domain::MerchantAccount,
+    time_range: api::TimeRange,
+) -> RouterResponse<api_models::refunds::RefundAggregateResponse> {
+    let db = state.store.as_ref();
+    let refund_status_with_count = db
+        .get_refund_status_with_count(merchant.get_id(), &time_range, merchant.storage_scheme)
+        .await
+        .change_context(errors::ApiErrorResponse::InternalServerError)
+        .attach_printable("Failed to find status count")?;
```
Comment (Chethan-rao) [inline crates/router/src/core/refunds.rs:1049]: Nit:

```suggestion
        .attach_printable("Failed to find refund status count")?;
```

**Review by Chethan-rao — APPROVED** [review]

**Review by ShivanshMathurJuspay — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Pull request merged**